File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaquePreditorDeBits.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Previsibilidade: treina um preditor por contexto (os k bits anteriores) na
 * primeira metade do keystream e mede a taxa de acerto na segunda metade.
 * Para um gerador sem memória/correlação local, a taxa fica em ~50% (chute).
 * Qualquer coisa acima disso indica que os bits carregam dependência
 * explorável - o embrião de um ataque de predição de estado.
 */
public class AtaquePreditorDeBits implements AtaqueInterface {
    private final int tamanho;
    private final int contexto;
    private final double limiteTaxa;

    public AtaquePreditorDeBits() {
        this(16384, 8, 0.55);
    }

    public AtaquePreditorDeBits(int tamanho, int contexto, double limiteTaxa) {
        this.tamanho = tamanho;
        this.contexto = contexto;
        this.limiteTaxa = limiteTaxa;
    }

    @Override
    public String nome() {
        return "Previsibilidade de bits (preditor por contexto)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] ks = alvo.gerarKeystreamBruto(
                alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8),
                Util.randomBytes(alvo.tamanhoIv()),
                "enc",
                this.tamanho);
        if (ks == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        int[] bits = this.paraBits(ks);
        int n = bits.length;
        int k = this.contexto;
        int total = 1 << k;
        int metade = n / 2;

        int[] uns = new int[total];
        int[] cont = new int[total];
        for (int i = k; i < metade; i++) {
            int ctx = this.contextoDe(bits, i, k);
            uns[ctx] += bits[i];
            cont[ctx]++;
        }

        int acertos = 0;
        int testes = 0;
        for (int i = metade; i < n; i++) {
            int ctx = this.contextoDe(bits, i, k);
            if (cont[ctx] == 0) {
                continue;
            }
            int predicao = uns[ctx] * 2 > cont[ctx] ? 1 : 0;
            if (predicao == bits[i]) {
                acertos++;
            }
            testes++;
        }

        double taxa = testes > 0 ? (double) acertos / testes : 0.5;
        boolean vulneravel = taxa > this.limiteTaxa;

        Map<String, Object> dados = new HashMap<>();
        dados.put("taxa", taxa);
        dados.put("testes", testes);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.ALTA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "taxa de acerto=%.2f%% com contexto de %d bits (esperado ~50%%, limite=%.0f%%) sobre %d testes",
                        taxa * 100,
                        k,
                        this.limiteTaxa * 100,
                        testes),
                dados);
    }

    private int contextoDe(int[] bits, int i, int k) {
        int ctx = 0;
        for (int j = i - k; j < i; j++) {
            ctx = (ctx << 1) | bits[j];
        }
        return ctx;
    }

    private int[] paraBits(byte[] bytes) {
        int[] bits = new int[bytes.length * 8];
        int idx = 0;
        for (int i = 0; i < bytes.length; i++) {
            int b = bytes[i] & 0xFF;
            for (int j = 7; j >= 0; j--) {
                bits[idx++] = (b >> j) & 1;
            }
        }
        return bits;
    }
}

File: packages/go/internal/seguranca/ataques/ataque_preditor_de_bits.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesPreditorDeBits configura o AtaquePreditorDeBits.
type OpcoesPreditorDeBits struct {
	Tamanho    int
	Contexto   int
	LimiteTaxa float64
}

// AtaquePreditorDeBits treina um preditor por contexto (os k bits anteriores)
// na primeira metade do keystream e mede a taxa de acerto na segunda metade.
// Para um gerador sem memória/correlação local a taxa fica em ~50%.
type AtaquePreditorDeBits struct {
	tamanho    int
	contexto   int
	limiteTaxa float64
}

// NovoAtaquePreditorDeBits cria o ataque. Sem argumento, usa tamanho = 16384,
// contexto = 8 e limiteTaxa = 0.55.
func NovoAtaquePreditorDeBits(opcoes ...OpcoesPreditorDeBits) *AtaquePreditorDeBits {
	o := OpcoesPreditorDeBits{Tamanho: 16384, Contexto: 8, LimiteTaxa: 0.55}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaquePreditorDeBits{tamanho: o.Tamanho, contexto: o.Contexto, limiteTaxa: o.LimiteTaxa}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaquePreditorDeBits) Nome() string {
	return "Previsibilidade de bits (preditor por contexto)"
}

// Executar roda o ataque.
func (a *AtaquePreditorDeBits) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	ks := alvo.GerarKeystreamBruto([]byte(alvo.ChaveDeTeste()), seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.tamanho)
	bits := a.paraBits(ks)
	n := len(bits)
	k := a.contexto
	total := 1 << k
	metade := n / 2

	uns := make([]int, total)
	cont := make([]int, total)
	for i := k; i < metade; i++ {
		ctx := a.contextoDe(bits, i, k)
		uns[ctx] += bits[i]
		cont[ctx]++
	}

	acertos := 0
	testes := 0
	for i := metade; i < n; i++ {
		ctx := a.contextoDe(bits, i, k)
		if cont[ctx] == 0 {
			continue
		}
		predicao := 0
		if uns[ctx]*2 > cont[ctx] {
			predicao = 1
		}
		if predicao == bits[i] {
			acertos++
		}
		testes++
	}

	taxa := 0.5
	if testes > 0 {
		taxa = float64(acertos) / float64(testes)
	}
	vulneravel := taxa > a.limiteTaxa

	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevAlta
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"taxa de acerto=%.2f%% com contexto de %d bits (esperado ~50%%, limite=%.0f%%) sobre %d testes",
			taxa*100, k, a.limiteTaxa*100, testes,
		),
		Dados: map[string]any{"taxa": taxa, "testes": testes},
	}, nil
}

// contextoDe monta o contexto de k bits anteriores à posição i.
func (a *AtaquePreditorDeBits) contextoDe(bits []int, i, k int) int {
	ctx := 0
	for j := i - k; j < i; j++ {
		ctx = (ctx << 1) | bits[j]
	}
	return ctx
}

// paraBits converte bytes em bits, do mais significativo ao menos.
func (a *AtaquePreditorDeBits) paraBits(bytes []byte) []int {
	bits := make([]int, 0, len(bytes)*8)
	for i := 0; i < len(bytes); i++ {
		by := bytes[i]
		for b := 7; b >= 0; b-- {
			bits = append(bits, int((by>>b)&1))
		}
	}
	return bits
}
